zapai/core.go

```diff
@@ -0,0 +1,125 @@
+package zapai
+
+import (
+	"github.com/microsoft/ApplicationInsights-Go/appinsights"
+	"github.com/microsoft/ApplicationInsights-Go/appinsights/contracts"
+	"github.com/pkg/errors"
+	"go.uber.org/zap/zapcore"
+)
+
+var levelToSev = map[zapcore.Level]contracts.SeverityLevel{
+	zapcore.DebugLevel:  contracts.Verbose,
+	zapcore.InfoLevel:   contracts.Information,
+	zapcore.WarnLevel:   contracts.Warning,
+	zapcore.ErrorLevel:  contracts.Error,
+	zapcore.DPanicLevel: contracts.Critical,
+	zapcore.PanicLevel:  contracts.Critical,
+	zapcore.FatalLevel:  contracts.Critical,
+}
+
+var _ zapcore.Core = (*Core)(nil)
+
+// Core implements zapcore.Core for appinsights.
+// Core only knows how to write to, and should only be used with, a corresponding appinsights Sink.
+//
+// Internally, the Core implements Write(zapcore.Entry, []zapcore.Field) by building an appinsights.TraceTelemetry
+// out of the zap inputs, then serializing it to a []byte using a shared common gob.Encoder, and passes that to
+// the Sink.Write([]byte) (which knows how to decode it back in to a TraceTelemetry).
+type Core struct {
+	zapcore.LevelEnabler
+	enc          traceEncoder
+	fieldMappers map[string]fieldTagMapper
+	fields       []zapcore.Field
+	out          zapcore.WriteSyncer
+}
+
+// NewCore creates a new appinsights zap core. Should only be initialized using an appinsights Sink as the
+// zapcore.WriteSyncer argument - other sinks will not error but will also not produce meaningful output.
+func NewCore(le zapcore.LevelEnabler, out zapcore.WriteSyncer) *Core {
+	return &Core{
+		LevelEnabler: le,
+		enc:          newTraceEncoder(),
+		fieldMappers: make(map[string]fieldTagMapper),
+		out:          out,
+	}
+}
+
+func (c *Core) WithFieldMappers(fieldMappers ...map[string]string) *Core {
+	clone := c.clone()
+	for _, fieldMapper := range fieldMappers {
+		for field, tag := range fieldMapper {
+			field, tag := field, tag // anchor vars in the closure
```

Comment:
`field` is not closed over, does it need to be new var in this scope?

Reply:
updated